reconcile/terraform_cloudflare_dns.py

```diff
@@ -0,0 +1,292 @@
+import logging
+import sys
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Union,
+)
+
+from reconcile.gql_definitions.terraform_cloudflare_dns import (
+    terraform_cloudflare_zones,
+)
+from reconcile.gql_definitions.terraform_cloudflare_dns.terraform_cloudflare_zones import (
+    AWSAccountV1,
+    CloudflareAccountV1,
+    CloudflareDnsZoneQueryData,
+    CloudflareDnsZoneV1,
+)
+from reconcile.status import ExitCodes
+from reconcile.typed_queries.app_interface_vault_settings import (
+    get_app_interface_vault_settings,
+)
+from reconcile.utils import gql
+from reconcile.utils.defer import defer
+from reconcile.utils.exceptions import SecretIncompleteError
+from reconcile.utils.external_resources import ExternalResourceSpec
+from reconcile.utils.runtime.integration import (
+    DesiredStateShardConfig,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.secret_reader import (
+    SecretReaderBase,
+    create_secret_reader,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.terraform.config_client import TerraformConfigClientCollection
+from reconcile.utils.terraform_client import TerraformClient
+from reconcile.utils.terrascript.cloudflare_client import (
+    DEFAULT_CLOUDFLARE_ACCOUNT_2FA,
+    DEFAULT_CLOUDFLARE_ACCOUNT_TYPE,
+    CloudflareAccountConfig,
+    TerraformS3BackendConfig,
+    TerrascriptCloudflareClient,
+    create_cloudflare_terrascript,
+)
+
+DEFAULT_NAMESPACE: Mapping[str, Any] = {
+    "name": False,
+    "cluster": {"name": False},
+    "environment": {"name": False},
+    "app": {"name": False},
+}
+DEFAULT_PROVISIONER_PROVIDER = "cloudflare"
+DEFAULT_PROVIDER = "zone"
+DEFAULT_EXCLUDE_KEY = "account"
+
+
+class TerraformCloudflareDNSIntegration(QontractReconcileIntegration):
+    def __init__(self) -> None:
+        self.qontract_integration = "terraform_cloudflare_dns"
+        self.qontract_integration_version = make_semver(0, 1, 0)
+        self.qontract_tf_prefix = "qrtfcfdns"
+
+    @property
+    def name(self) -> str:
+        return self.qontract_integration.replace("_", "-")
+
+    def create_backend_config(
```

Comment:
What is the purpose of the `TerraformCloudflareDNSIntegration`? What does it do? Does it need this method and others that were previously functions?

Reply:
This is part of https://github.com/app-sre/qontract-reconcile/pull/3134/files#diff-dc5ee8a177d8eecdd1e8f2adda7b5dfc0481a16aa56bf830a8e26558439bef5f 
Instead of adding integrations as module, they are now getting added as classes that implements TerraformCloudflareDNSIntegration. I'm not sure I'm following the second part of your questions. The methods are the same as they were functions.

Reply:
So, the purpose of `TerraformCloudflareDNSIntegration` is to implement `QontractReconcileIntegration` and standardize some aspects of how we run integrations. My question is why this class has to know anything about creating backend configs, configuring accounts, building clients, etc? Basically what I'm getting at is the smaller the classes are (within reason), and the less that they do, the easier they will generally be to understand and test.

Reply:
Talked offline. Agree with the approach as it breaks the logic up. It will produce more specialized classes, and the functions stays in the module will facilitate later iteration that abstract those functionalities.